Windows 10 + Visual Studio 2022 setup

I finally got Visual Studio set up where KQ compiles *and* runs. I've pushed these changes to [dev/onlinecop/vs2022](https://github.com/OnlineCop/kq-fork/tree/dev/onlinecop/vs2022).

1. I believe I had a bug in one of my VS2019 plugins (unrelated to KQ), causing it to crash.
2. I downloaded Visual Studio Community 2022 using the Visual Studio Installer, and selected:
    - `Desktop development with C++`
    - `Game development with Unity` (probably not required for your setup)
    - `Game development with C++` (may also not be required for your setup)
    - `Linux and embedded development with C++` (probably not required for your setup)
3. Taking advice on #74, as well as the updates on the [EngineRework](https://github.com/OnlineCop/kq-fork/wiki/EngineRework#building-on-windows-with-new-sdl-branch-on-road_to_10) wiki page, I made the following system and repo changes:
    1. I installed vcpkg to `C:\Users\user\source\repos\vcpkg` according to its [Getting Started](https://vcpkg.io/en/getting-started.html) guide for Windows.
    2. I added @z9484's `vcpkg.json` file to the project:
        ```json
        {
            "name": "kq",
            "version-string": "0.99",
            "dependencies": [
                "pkgconf",
                "sdl2",
                "sdl2-mixer",
                "lua",
                "libpng",
                "zlib",
                "tinyxml2"
            ]
        }
        ```
    3. I only had to modify one line in `CMakeSettings.json`:
        - From: `"cmakeToolchain": "C:/Users/`**`Peter`**`/Source/Repos/vcpkg/scripts/buildsystems/vcpkg.cmake"`
        - To: `"cmakeToolchain": "C:/Users/`**`user`**`/Source/Repos/vcpkg/scripts/buildsystems/vcpkg.cmake"`
    4. I removed one line in `CMakeLists.txt`:
        - Line 2: `include(FindPkgConfig)`
    5. I ran  `vcpkg.exe install` from within the `kq-fork/` root directory, which installed packages under `vcpkg_installed/`.
    6. Within VS2022, I opened the `C:\Users\user\Source\Repos\kq-fork` directory:
        1. I selected Build->Build All
        2. When the build completed, from the terminal in the `kq-fork` root, I ran: `out/build/x64-Debug/kq-fork.exe`
        3. The game runs!

I plan to modify this Issue as I create a fresh Windows user account and reinstall the various components, to ensure some undocumented steps didn't come into play here.

This issue was also created to facilitate discussions for anyone wishing to ask/answer questions specific to getting KQ to run on Windows + VS2022 (the wiki doesn't appear to have an easy way to discuss stuff).

Let's see if I've done this correctly:

1. Using Git Bash (so I don't accidentally trigger the WSL2 layer) in `/c/Users/user/source/repos/`:
    ```bash
    git clone -b road_to_1.0 git@github.com:OnlineCop/kq-fork.git kq-fork-2
    ```
2. Within `/c/Users/user/source/repos/kq-fork-2/`, pull in PR #76:
    ```bash
    ~/source/repos/kq-fork-2 [road_to_1.0]$ git fetch origin pull/76/head:pedro-pr-76
    remote: Enumerating objects: 4, done.
    remote: Counting objects: 100% (4/4), done.
    remote: Total 4 (delta 3), reused 3 (delta 3), pack-reused 0
    Unpacking objects: 100% (4/4), 798 bytes | 31.00 KiB/s, done.
    From github.com:OnlineCop/kq-fork
     * [new ref]         refs/pull/76/head -> pedro-pr-76
    ```
3. Switch to that branch:
    ```bash
    ~/source/repos/kq-fork-2 [road_to_1.0]$ git checkout pedro-pr-76
    Switched to branch 'pedro-pr-76'

    ~/source/repos/kq-fork-2 [pedro-pr-76]$
    ```
4. Run cmake, assuming `vcpkg` is located in a sibling directory:
    ```bash
    ~/source/repos/kq-fork-2 [pedro-pr-76]$ cmake -B build -S . -DKQ_DATADIR="$PWD" -DCMAKE_TOOLCHAIN_FILE="$PWD"/../vcpkg/scripts/buildsystems/vcpkg.cmake
    ```
    (I realize the `"$PWD"` may be overkill, but it was giving me odd runtime behavior when I used `.` as the current directory in earlier runs.)

    When I do this, it fails to find LUA_LIBRARIES (it might fail to find the other libraries after that as well, but it always stops on the first missing library):

    ```
    -- Building for: Visual Studio 17 2022
    -- Selecting Windows SDK version 10.0.19041.0 to target Windows 10.0.19044.
    -- The C compiler identification is MSVC 19.32.31329.0
    -- The CXX compiler identification is MSVC 19.32.31329.0
    -- Detecting C compiler ABI info
    -- Detecting C compiler ABI info - done
    -- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.32.31326/bin/Hostx64/x64/cl.exe - skipped
    -- Detecting C compile features
    -- Detecting C compile features - done
    -- Detecting CXX compiler ABI info
    -- Detecting CXX compiler ABI info - done
    -- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.32.31326/bin/Hostx64/x64/cl.exe - skipped
    -- Detecting CXX compile features
    -- Detecting CXX compile features - done
    CMake Error at C:/Program Files/CMake/share/cmake-3.22/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
      Could NOT find Lua (missing: LUA_LIBRARIES LUA_INCLUDE_DIR)
    Call Stack (most recent call first):
      C:/Program Files/CMake/share/cmake-3.22/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
      C:/Program Files/CMake/share/cmake-3.22/Modules/FindLua.cmake:233 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
      C:/Users/user/source/repos/vcpkg/scripts/buildsystems/vcpkg.cmake:824 (_find_package)
      CMakeLists.txt:9 (find_package)


    -- Configuring incomplete, errors occurred!
    See also "C:/Users/user/source/repos/kq-fork-2/build/CMakeFiles/CMakeOutput.log".
    ```
5. Re-running it but with the full path instead of `"$PWD"` for the sibling directory yields pretty much the same:
    ```bash
    ~/source/repos/kq-fork-2 [pedro-pr-76]$ cmake -B build -S . -DKQ_DATADIR="$PWD" -DCMAKE_TOOLCHAIN_FILE=/c/Users/user/source/repos/vcpkg/scripts/buildsystems/vcpkg.cmake
    -- Selecting Windows SDK version 10.0.19041.0 to target Windows 10.0.19044.
    CMake Error at C:/Program Files/CMake/share/cmake-3.22/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
      Could NOT find Lua (missing: LUA_LIBRARIES LUA_INCLUDE_DIR)
    Call Stack (most recent call first):
      C:/Program Files/CMake/share/cmake-3.22/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
      C:/Program Files/CMake/share/cmake-3.22/Modules/FindLua.cmake:233 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
      C:/Users/user/source/repos/vcpkg/scripts/buildsystems/vcpkg.cmake:824 (_find_package)
      CMakeLists.txt:9 (find_package)


    -- Configuring incomplete, errors occurred!
    See also "C:/Users/user/source/repos/kq-fork-2/build/CMakeFiles/CMakeOutput.log".
    ````

From `vcpkg list`, I have:

    $ vcpkg list
    libpng:x86-windows                                 1.6.37#17        libpng is a library implementing an interface fo...
    lua:x86-windows                                    5.4.4#2          A powerful, fast, lightweight, embeddable script...
    pkgconf:x86-windows                                1.8.0#3          pkgconf is a program which helps to configure co...
    sdl2-mixer:x86-windows                             2.0.4#16         Multi-channel audio mixer library for SDL.
    sdl2-mixer[nativemidi]:x86-windows                                  Support for MIDI audio format on Windows and macOS.
    sdl2:x64-windows                                   2.0.22#1         Simple DirectMedia Layer is a cross-platform dev...
    sdl2:x86-windows                                   2.0.22#1         Simple DirectMedia Layer is a cross-platform dev...
    tinyxml2:x86-windows                               9.0.0#1          A simple, small, efficient, C++ XML parser
    vcpkg-cmake-config:x64-windows                     2022-02-06#1
    vcpkg-cmake:x64-windows                            2022-05-10#1
    vcpkg-tool-meson:x64-windows                       0.62.1           Meson build system
    zlib:x86-windows                                   1.2.12#1         A compression library

That's probably my issue then: lua is installed as x86 for some reason, and when I run cmake in the directory, it claims it can't find it:

    $ cmake -S . -B build -DCMAKE_TOOLCHAIN_FILE=/c/Users/user/source/repos/vcpkg/scripts/buildsystems/vcpkg.cmake
    -- Selecting Windows SDK version 10.0.19041.0 to target Windows 10.0.19044.
    CMake Error at C:/Program Files/CMake/share/cmake-3.22/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
      Could NOT find Lua (missing: LUA_LIBRARIES LUA_INCLUDE_DIR)
    Call Stack (most recent call first):
      C:/Program Files/CMake/share/cmake-3.22/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
      C:/Program Files/CMake/share/cmake-3.22/Modules/FindLua.cmake:233 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
      C:/Users/user/source/repos/vcpkg/scripts/buildsystems/vcpkg.cmake:824 (_find_package)
      CMakeLists.txt:9 (find_package)

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,29 +1,41 @@
 cmake_minimum_required(VERSION 3.18)
-include(FindPkgConfig)
 
 project(kq-fork)
 
 set (CMAKE_MODULE_PATH "${CMAKE_MODULE_PATH}"
   "${CMAKE_CURRENT_SOURCE_DIR}")
 
-find_package(SDL2 CONFIG REQUIRED)
+find_package(SDL2 REQUIRED)
 find_package(Lua REQUIRED)
 find_package(PNG REQUIRED)
-find_package(tinyxml2 CONFIG REQUIRED)
-if(WIN32)
-find_package(sdl2-mixer CONFIG REQUIRED)
-set(SDL2_LIBRARIES SDL2::SDL2 SDL2::SDL2main SDL2::SDL2_mixer)
+find_package(tinyxml2)
+find_package(sdl2-mixer)
+
+# homebrew doesn't seem to have a FindXXX module for tinyxml2 and SDL_mixer, but vcpkg does,
+# therefore if find_package doesn't work, fall back to pkg-config. There is (usually) no
+# pkg-config on windows so it's included conditionally.
+if(tinyxml2_FOUND)
+  set(TINYXML2_LINK_LIBRARIES tinyxml2::tinyxml2)
 else()
-pkg_check_modules(SDL2_MIXER REQUIRED sdl2_mixer)
-set(SDL2_LIBRARIES SDL2::SDL2 SDL2::SDL2main ${SDL2_MIXER_LINK_LIBRARIES})
+  find_package(PkgConfig REQUIRED)
+  pkg_check_modules(TINYXML2 REQUIRED tinyxml2)
 endif()
-configure_file(config.cmake.h makeconfig.h)
+
+if(sdl2-mixer_FOUND)
+  set(SDL2_LIBRARIES SDL2::SDL2 SDL2::SDL2main SDL2::SDL2_mixer)
+else()
+  find_package(PkgConfig REQUIRED)
+  pkg_check_modules(SDL2_MIXER REQUIRED SDL2_mixer)
+  set(SDL2_LIBRARIES SDL2::SDL2 SDL2::SDL2main ${SDL2_MIXER_LINK_LIBRARIES})
+endif()
+
 include_directories(${CMAKE_BINARY_DIR}) # for makeconfig.h
 include_directories( ${LUA_INCLUDE_DIR} ${TINYXML2_INCLUDE_DIR} ${PNG_INCLUDE_DIRS} ${ZLIB_INCLUDE_DIRS} ${SDL2_INCLUDE_DIRS} ${SDL2_mixer_INCLUDE_DIRS} ${SDL2_MIXER_INCLUDE_DIRS})
 add_definitions(${PNG_DEFINITIONS})
 set(KQ_DEBUGGING 1 CACHE BOOL "Features for debugging the game")
 set(KQ_CHEATS 1 CACHE BOOL "Features for allowing Lua script cheats in game")
-set(KQ_DATADIR "." CACHE STRING "Where the data dirs are going to be installed")
+set(KQ_DATADIR "." CACHE PATH "Where the data dirs are going to be installed")
+configure_file(config.cmake.h makeconfig.h)
 
 set(kq-fork_SRCS
 	src/anim_sequence.cpp
@@ -65,10 +77,9 @@ set(kq-fork_SRCS
 	src/shopmenu.cpp
 	src/tiledmap.cpp
 	src/timing.cpp
-	$<$<PLATFORM_ID:Unix,Darwin>:src/unix.cpp>
+	$<$<PLATFORM_ID:Linux,Darwin>:src/unix.cpp>
 	$<$<PLATFORM_ID:Windows>:src/win.cpp>
 )
-set(TINYXML2_LIBRARIES tinyxml2::tinyxml2)
 set(PNG_LIBRARIES PNG::PNG)
 
 include_directories("${CMAKE_CURRENT_SOURCE_DIR}/include")
@@ -79,6 +90,5 @@ target_compile_features(kq-fork PUBLIC cxx_std_17)
 target_link_libraries(kq-fork
 	${LUA_LIBRARIES}
 	${SDL2_LIBRARIES}
-	${TINYXML2_LIBRARIES}
-	${M_LIB}
+	${TINYXML2_LINK_LIBRARIES}
 	${PNG_LIBRARIES})

--- a/CMakeSettings.json
+++ b/CMakeSettings.json
@@ -9,8 +9,7 @@
       "installRoot": "${projectDir}\\out\\install\\${name}",
       "cmakeCommandArgs": "",
       "buildCommandArgs": "",
-      "ctestCommandArgs": "",
-      "cmakeToolchain": "C:/Users/Peter/Source/Repos/vcpkg/scripts/buildsystems/vcpkg.cmake"
+      "ctestCommandArgs": ""
     }
   ]
 }

